sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,912 @@
+//! @brief Solana Rust-based BigNumber for bpf-programs
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use std::fmt;
+
+#[repr(C)]
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    AbiExample,
+)]
+
+/// BigNumber supports integers greater than
+/// 64 bits for BPF programs
+pub struct BigNumber {
+    value: Vec<u8>,
+    is_negative: bool,
+}
+
+impl BigNumber {
+    /// Returns a BigNumber with initial value of 0
+    pub fn new() -> Self {
+        Self {
+            is_negative: false,
+            value: vec![0u8],
+        }
+    }
+
+    /// Returns the size, in bytes, of BigNum. Typically used in
+    /// assessing buffer size needed to fetch the bytes for serialization, etc.
+    pub fn size_in_bytes(&self) -> usize {
+        self.value.len()
+    }
+
+    /// Returns a BigNumber with initial value set to a u32 value
+    pub fn from_u32(val: u32) -> Result<Self, u64> {
+        if val == 0 {
+            Ok(Self::new())
+        } else {
+            #[cfg(not(target_arch = "bpf"))]
+            {
+                match openssl::bn::BigNum::from_u32(val) {
+                    Ok(bn) => Ok(Self {
+                        is_negative: false,
+                        value: bn.to_vec(),
+                    }),
+                    Err(_) => Err(0x1),
+                }
+            }
+            #[cfg(target_arch = "bpf")]
+            {
+                extern "C" {
+                    fn sol_bignum_from_u32(
+                        bytes_ptr: *mut u8,
+                        bytes_len: *mut u64,
+                        val: u64,
+                    ) -> u64;
+                }
+                let mut value = Vec::<u8>::with_capacity(4);
```

Comment:
Could this be `let mut value = &[0u8; 4]` to avoid the allocation?

Reply:
the issue is that I need a buffer sized on the program side (hence the heap alloc) as I do not know "a priori" how big the resulting value will be. For example from_u32 I moved to a fixed 4 byte for going into the syscall but the actual result may be 1 to 4 bytes